File: DevEnvironment/Base/src/main/java/base/core/WindowsPath.java
package base.core;

import base.interfaces.IPathBuilder;
import java.util.Stack;

/**
 * The WindowsPath class implements a windows path.
 */
public final class WindowsPath extends AbstractPath {
    public static final char SEPARATOR = '\\';

    public static final String DIRECTORY_SEPARATOR = String.valueOf(SEPARATOR);
    public static final String DIRECTORY_SEPARATOR_REGEX = "\\\\";
    public static final String CURRENT_DIRECTORY = Paths.CURRENT_DIRECTORY;
    public static final String PARENT_DIRECTORY = Paths.PARENT_DIRECTORY;

    /**
     * The WindowsPath constructor.
     */
    public WindowsPath(String path) {
        super(path);
    }

    /**
     * Gets the directory section of the path.
     */
    public static String getDirectory(String path) {
        return Paths.getDirectory(path, WindowsPath.DIRECTORY_SEPARATOR);
    }

    /**
     * Makes an absolute path.
     */
    @Override
    protected String makeAbsolutePath(String path) {
        return WindowsPath.makeAbsolute(path);
    }

    /**
     * Makes a canonical path.
     */
    @Override
    protected String makeCanonicalPath(String path) {
        return WindowsPath.makeCanonical(path);
    }

    /**
     * Validates a path.
     */
    @Override
    protected void validatePath(String path) {
        WindowsPath.validate(path);
    }

    /**
     * Validates a unix path.
     */
    public static void validate(String path) {
        Conditions.validateStringNotNullOrEmpty(
            path,
            "The windows path.");
    }

    /**
     * Makes an absolute path.
     */
    private static String makeAbsolute(String path) {
        return makeCanonical(path);
    }

    /**
     * Makes a canonical path.
     */
    private static String makeCanonical(String path) {
        String[] components = path.split(WindowsPath.DIRECTORY_SEPARATOR_REGEX);

        Stack<String> stack = new Stack<>();

        for (String component : components) {
            if (component.equals(WindowsPath.PARENT_DIRECTORY)) {
                stack.pop();
            }
            else if (!component.isEmpty() &&
                     !component.equals(WindowsPath.CURRENT_DIRECTORY)) {

                stack.push(component);
            }
        }

        IPathBuilder pathBuilder = new PathBuilder();

        for (String component : stack) {
            pathBuilder.addComponent(component);
        }

        return pathBuilder.build();
    }
}

File: SystemCore/Base/core/WindowsPath.cpp
#include "WindowsPath.h"
#include "PathBuilder.h"
#include "Strings.h"
#include "InvalidArgumentException.h"
#include <stack>
#include <experimental/filesystem>

using namespace base;
using namespace base::environment::windows;
const std::wstring::value_type WindowsPath::SEPARATOR = L'\\';
const std::wstring WindowsPath::DIRECTORY_SEPARATOR = std::wstring(1, SEPARATOR);
const std::wstring::value_type WindowsPath::EXTENSION_SEPARATOR = L'.';
const std::wstring WindowsPath::CURRENT_DIRECTORY = L".";
const std::wstring WindowsPath::PARENT_DIRECTORY = L"..";

/**
 * Creates a new windows path.
 */
IPathSharedPtr WindowsPath::Make(const std::wstring& path)
{
    return std::make_shared<WindowsPath>(path);
}

/**
 * The WindowsPath constructor.
 */
WindowsPath::WindowsPath(const std::wstring& path) :
    AbstractPath(
        path,
        DIRECTORY_SEPARATOR,
        EXTENSION_SEPARATOR)
{
}

/**
 * The WindowsPath destructor.
 */
WindowsPath::~WindowsPath()
{
}

/**
 * Gets seperator.
 */
std::wstring::value_type WindowsPath::Seperator()
{
    return SEPARATOR;
}

/**
 * Gets directory seperator.
 */
std::wstring WindowsPath::DirectorySeperator()
{
    return DIRECTORY_SEPARATOR;
}

/**
 * Gets current directory.
 */
std::wstring WindowsPath::CurrentDirectory()
{
    return CURRENT_DIRECTORY;
}

/**
 * Gets parent directory.
 */
std::wstring WindowsPath::ParentDirectory()
{
    return PARENT_DIRECTORY;
}

/**
 * Validates a path.
 */
void WindowsPath::ValidatePath(const std::wstring& path) const
{
    if (path.empty()) {
        throw InvalidArgumentException(L"The input path can not be empty.");
    }
}

/**
 * Makes an absolute path.
 */
std::wstring WindowsPath::MakeAbsolutePath(const std::wstring& path) const
{
    return MakeAbsolute(path);
}

/**
 * Makes a canonical path.
 */
std::wstring WindowsPath::MakeCanonicalPath(const std::wstring& path) const
{
    return MakeCanonical(path);
}

/**
 * Makes an absolute path.
 */
std::wstring WindowsPath::MakeAbsolute(const std::wstring& path)
{
    return MakeCanonical(path);
}

/**
 * Makes a canonical path.
 */
std::wstring WindowsPath::MakeCanonical(const std::wstring& path)
{
    //
    // Split tokens by directory seperator...
    //
    std::vector<std::wstring> tokens;
    Strings::SplitString(path, DirectorySeperator(), tokens);

    //
    // Process the tokens with a stack...
    //
    std::stack<std::wstring> stack;

    for (const std::wstring& token : tokens) {
        if (token == ParentDirectory())
        {
            stack.pop();
        }
        else if (!token.empty() && token != CurrentDirectory())
        {
            stack.push(token);
        }
    }

    //
    // Create the list of finalize component...
    //
    std::list<std::wstring> components;

    while (!stack.empty()) {
        std::wstring component = stack.top();
        stack.pop();
        components.push_front(component);
    }

    //
    // Assemble the path...
    //
    PathBuilder pathBuilder;

    for (const std::wstring& component : components) {
        pathBuilder.AddComponent(component);
    }

    return pathBuilder.Build();
}

/**
 * Checks whether a path is absolute.
 */
bool WindowsPath::IsAbsolute(const std::wstring& path) const
{
    return !IsRelative(path);
}

/**
 * Checks whether a path is relative.
 */
bool WindowsPath::IsRelative(const std::wstring& path) const
{
    std::experimental::filesystem::path systemPath(path);
    return systemPath.is_relative();
}
